db/log_reader.cc

```diff
@@ -302,25 +332,61 @@ unsigned int Reader::ReadPhysicalRecord(Slice* result,
       // Skip zero length record without reporting any drops since
       // such records are produced by the mmap based writing code in
       // env_posix.cc that preallocates file regions.
+      //
       // NOTE: this should never happen in DB written by new RocksDB versions,
       // since we turn off mmap writes to manifest and log files
       buffer_.clear();
+      if (recycled_ &&
+	  wal_recovery_mode != WALRecoveryMode::kSkipAnyCorruptedRecords) {
+	return kEof;
+      }
       return kBadRecord;
     }
 
     // Check crc
     if (checksum_) {
-      uint32_t expected_crc = crc32c::Unmask(DecodeFixed32(header));
-      uint32_t actual_crc = crc32c::Value(header + 6, 1 + length);
+      uint32_t expected_crc, actual_crc;
+      bool try_recycled = false;
+      if (!recycled_) {
+	// Simple CRC (no log number)
+	expected_crc = crc32c::Unmask(DecodeFixed32(header));
+	actual_crc = crc32c::Value(header + 6, 1 + length);
+	if (last_record_offset_ == 0 && actual_crc != expected_crc) {
+	  try_recycled = true;
+	}
+      }
+      if (recycled_ || try_recycled) {
+	uint32_t stored_crc = crc32c::Unmask(DecodeFixed32(header));
+	expected_crc = stored_crc ^ static_cast<uint32_t>(log_number_);
+	actual_crc = crc32c::Value(header + 6, 1 + length);
+	if (try_recycled && db_options_)
+	  Log(InfoLogLevel::INFO_LEVEL, db_options_->info_log,
```

Comment:
Is it the only place we use db_options_? If that is the case, I suggest we pass db_options_->info_log directly.